[WIP] Additional features/customizations

thechrisnixon opened the pull request:
@Sharlaan started adding some features/customizations to the library as I've been using the library in a prod app. Let me know what you think of the changes. Still need to do some testing, add documentation, etc.

Allow optional/customizable select-all

Allow inner and outer styles on selections presenter

Allow custom footer height, autocomplete height & no match found height

Allow selections renderer to always be displayed and make the floating
label stay shrunken

TarikHuber committed:
Merge pull request #1 from Sharlaan/master

Merge

Sharlaan committed:
first commit

thechrisnixon committed:
Allow optional/customizable select-all

Allow inner and outer styles on selections presenter

Allow custom footer height, autocomplete height & no match found height

Allow selections renderer to always be displayed and make the floating
label stay shrunken

Sharlaan reviewed (commented):
Rename `alwaysDisplaySelectionsRenderer` with `floatingLabelFixed` to stay consistent with Material-UI naming.

thechrisnixon committed:
Rename prop & add autowidth to popover

s/alwaysDisplaySelectionsRenderer/floatingLabelFixed
Add autoWidth prop for popover

thechrisnixon committed:
Don't use autoWidth and use menuStyle width as popover width when passed as prop

Sharlaan commented on src/SuperSelectField.js:
```diff
@@ -337,8 +350,18 @@ class SelectField extends Component {
+    if (allowSelectAll) {
+      const selectAllItemIndex = selectedItems.findIndex(item => item.value === selectAllItem.value);
+      if (selectAllItemIndex !== -1) {
+        selectedItems = [
```
You are mutating the state here, best practice is to first make a copy then play with it.

Also i don't understand what are you trying to achieve with these consecutive slices ?
Exclude the `selectAllItem` ? if yes, just use `.filter()`

thechrisnixon replied:
I'll fix this soon, along with using `const` (comment below)

Sharlaan commented on src/SuperSelectField.js:
```diff
@@ -337,8 +350,18 @@ class SelectField extends Component {
-    const { onChange, name } = this.props
-    onChange(this.state.selectedItems, name)
+    const { onChange, name, allowSelectAll, selectAllItem } = this.props
+    let { selectedItems } = this.state
```
Should use `const` to make sure to never mutate the state.

Sharlaan commented on src/SuperSelectField.js:
```diff
@@ -406,10 +429,35 @@ class SelectField extends Component {
-        : selectedItems.concat(selectedItem)
+      const { allowSelectAll, selectAllItem } = this.props;
+      let updatedValues;
+      if (allowSelectAll && selectedItem.value === selectAllItem.value) {
```
What is `selectAllItem` ? Can you document it in Readme.md please ? especially its value property.

Is it a bool ? if not, can you give a usecase please.

thechrisnixon replied:
It is an object of the format `{ value: 'all-superSelectField', label: 'All' }`. It is in the PropTypes for the select field. I figured it _could_ be possible someone might already have an existing value that is `all`, so wanted to give them the ability to change it if necessary.

Is there a better name for it?

TarikHuber committed:
Merge pull request #2 from Sharlaan/master

Preparing for v1.6.0 :

TarikHuber committed:
Removed PropTypes and createReactClass warnings

Sharlaan committed:
Merge pull request #36 from TarikHuber/master

Removed PropTypes and createReactClass warnings

thechrisnixon committed:
Add classNames lib and dont mutate state (duh!)

* Allow className to be added to root div (solve focus outline)
* Don't mutate state, in this case, `selectedItems`. Filter out the
select all item when select all is enabled

thechrisnixon commented:
@Sharlaan I rebased my branch off of master, would you please do that for your branch too (if you're still working off of it)? Otherwise I can just re-point to `master`

Sharlaan commented:
i forgot to tell you this:
would it be possible to recreate this PR with your own branch named, for exemple:
`dasnixon-selectall`
Also ideally should make 1 PR per feature, not 1 PR with 4-5 different packed features. Would be much easier to debug/accept/merge.

This branch `sharlaan-footer-renderer` is about some other feature totally offtopic

thechrisnixon commented:
@Sharlaan will do, thanks for the feedback.
